src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala

```diff
@@ -0,0 +1,209 @@
+  }
+
+  def withTasksAndDeployments(
+    appTasks: Seq[EnrichedTask], healthCounts: HealthCounts,
+    runningDeployments: Seq[DeploymentPlan]): V2AppDefinition.WithTasksAndDeployments =
+    new V2AppDefinition.WithTasksAndDeployments(appTasks, healthCounts, runningDeployments, this)
+
+  def withTasksAndDeploymentsAndFailures(
+    appTasks: Seq[EnrichedTask], healthCounts: HealthCounts,
+    runningDeployments: Seq[DeploymentPlan],
+    taskFailure: Option[TaskFailure]): V2AppDefinition.WithTasksAndDeploymentsAndTaskFailures =
+    new V2AppDefinition.WithTasksAndDeploymentsAndTaskFailures(
+      appTasks, healthCounts,
+      runningDeployments, taskFailure, this
+    )
+}
+
+object V2AppDefinition {
+
+  def apply(app: AppDefinition): V2AppDefinition =
+    V2AppDefinition(
+      app.id, app.cmd, app.args, app.user, app.env, app.instances, app.cpus,
+      app.mem, app.disk, app.executor, app.constraints, app.uris,
+      app.storeUrls, app.ports, app.requirePorts, app.backoff,
+      app.backoffFactor, app.maxLaunchDelay, app.container,
+      app.healthChecks, app.dependencies, app.upgradeStrategy,
+      app.labels, app.acceptedResourceRoles, app.version)
+
+  protected[marathon] class WithTaskCountsAndDeployments(
+    appTasks: Seq[EnrichedTask],
+    healthCounts: HealthCounts,
+    runningDeployments: Seq[DeploymentPlan],
+    private val app: V2AppDefinition)
+      extends V2AppDefinition(
+        app.id, app.cmd, app.args, app.user, app.env, app.instances, app.cpus,
+        app.mem, app.disk, app.executor, app.constraints, app.uris,
+        app.storeUrls, app.ports, app.requirePorts, app.backoff,
+        app.backoffFactor, app.maxLaunchDelay, app.container,
+        app.healthChecks, app.dependencies, app.upgradeStrategy,
+        app.labels, app.acceptedResourceRoles, app.version) {
+
+    /**
+      * Snapshot of the number of staged (but not running) tasks
+      * for this app
+      */
+    @JsonProperty
+    val tasksStaged: Int = appTasks.count { eTask =>
+      eTask.task.getStagedAt != 0 && eTask.task.getStartedAt == 0
+    }
+
+    /**
+      * Snapshot of the number of running tasks for this app
+      */
+    @JsonProperty
+    val tasksRunning: Int = appTasks.count { eTask =>
+      eTask.task.hasStatus &&
+        eTask.task.getStatus.getState == mesos.TaskState.TASK_RUNNING
+    }
+
+    /**
+      * Snapshot of the number of healthy tasks for this app
+      */
+    @JsonProperty
+    val tasksHealthy: Int = healthCounts.healthy
+
+    /**
+      * Snapshot of the number of unhealthy tasks for this app
+      */
+    @JsonProperty
+    val tasksUnhealthy: Int = healthCounts.unhealthy
+
+    /**
+      * Snapshot of the running deployments that affect this app
+      */
+    @JsonProperty
+    def deployments: Seq[Identifiable] = {
+      runningDeployments.collect {
+        case plan: Any if plan.affectedApplicationIds contains app.id => Identifiable(plan.id)
```

Comment:
Why did you change this from `plan: DeploymentPlan` to `plan: Any`? To be honest, I am surprised that this compiles.

Not related to this PR but out of curiosity: Why are we passing the runningDeployments to all AppDefinitions and filter them afterwards?